File: src/main/ruby/g0001_0100/s0045_jump_game_ii/solution.rb
# #Medium #Top_100_Liked_Questions #Array #Dynamic_Programming #Greedy
# #Algorithm_II_Day_13_Dynamic_Programming #Dynamic_Programming_I_Day_4
# #Big_O_Time_O(n)_Space_O(1) #2023_11_19_Time_68_ms_(75.23%)_Space_211.9_MB_(79.82%)

# @param {Integer[]} nums
# @return {Integer}
def jump(nums)
  length = 0
  max_length = 0
  min_jump = 0

  (0..nums.length - 2).each do |i|
    length -= 1
    max_length -= 1
    max_length = [max_length, nums[i]].max

    if length <= 0
      length = max_length
      min_jump += 1
    end

    return min_jump if length >= nums.length - i - 1
  end

  min_jump
end

File: src/main/scala/g0001_0100/s0045_jump_game_ii/Solution.scala
package g0001_0100.s0045_jump_game_ii

// #Medium #Top_100_Liked_Questions #Array #Dynamic_Programming #Greedy
// #Algorithm_II_Day_13_Dynamic_Programming #Dynamic_Programming_I_Day_4
// #Big_O_Time_O(n)_Space_O(1) #2024_06_01_Time_653_ms_(83.87%)_Space_58.6_MB_(87.10%)

object Solution {
    def jump(nums: Array[Int]): Int = {
        var length = 0
        var maxLength = 0
        var minJump = 0

        var result = -1
        for (i <- 0 until nums.length - 1 if result == -1) {
            length -= 1
            maxLength -= 1
            maxLength = math.max(maxLength, nums(i))

            if (length <= 0) {
                length = maxLength
                minJump += 1
            }

            if (length >= nums.length - i - 1) {
                result = minJump
            }
        }

        if (result == -1) minJump else result
    }
}
